docs/proposals/kubectl-apply-purge-from-namespace.md

```diff
@@ -0,0 +1,102 @@
+
+All resources referred to must be explicitly labelled with a non-default namespace in order for the `--purge-from-namespace` argument to succeed.
+
+This flag will require `--namespace` not to be set.
+All referenced resources must be explicitly defined as being in the same namespace, set in each of the resource definitions.
+
+The goal here is to avoid a negative user experience where the user accidentally deletes all the resources in some other namespace just because they accidentally refer to (e.g.) a single resource in some other namespace.
+
+## Managing Add-ons
+
+Let's first consider the add-ons use case.
+
+There exists a `/etc/kubernetes/addons` directory with some number of files. All resources declared in those files belong to the same namespace (`kube-system`), which is set explicitly in each of the resources.
+
+When `kubectl apply --purge-from-namespace -f /etc/kubernetes/addons` runs for the first time all the resources will be created.
+
+Eventually some new files appear in `/etc/kubernetes/addons` and some are modified.
+
+Subsequent invocations of `kubectl apply --purge-from-namespace -f /etc/kubernetes/addons` result in new resources being created and modified resources being updated.
+
+When user deletes some of the files or resources within files from `/etc/kubernetes/addons`, `kubectl apply --purge-from-namespace` would ensure the resources that were referenced in the deleted files are deleted, based on the assumption that `/etc/kubernetes/addons` represents all resources that are supposed to be present in `kube-system` namespace.
+
+
+## Managing Any Other Application
+
+It should be possible to apply the same approach to any other applications, with the assumption that single application occupies a namespace of its own.
+
+The usage may look like this:
+
+```
+kubectl apply --purge-from-namespace -f https://example.com/myapp1.json
+```
+
+Given resources in `myapp1.json` set their namespaces explicitly, resource that are no longer in this file would be purged, if no longer defined.
+
+```
+kubectl apply --purge-from-namespace -f https://example.com/myapp-part1.json -f https://example.com/myapp-part2.json
+```
+
+If `myapp1.json` and `myapp2.json` are in the same namespaces and it is explicitly defined, removing either of the URLs from the invocations arguments will result in all the resources that URL has defined to be purged.
+
+## Flag Discovery User Experience
+
+As stated above, the `--purge-from-namespace` flag shouldn't be enabled by default, however it should be discoverable.
+
+So, if a `kubectl apply` command is run without `--purge-from-namespace`, and the referenced files are all in the same namespace, then the command could output something like:
+
+```
+$ kubectl apply -f resources.yaml
+Updating ServiceA...
```

Comment:
That would be useful. 

The precise syntax should follow the output convention of other commands, including both the resource type and resource name in a form that could be copy+pasted to other commands verbatim, and should support `-o name` for scripting use.